### PR TITLE
Ensure orderly shutdown of ssl socket

### PR DESCRIPTION
**Pull request by jepler**
A crash would occur if an SSL socket was not shut down before `gc_deinit()`.

I do not fully understand the root cause, but some object deinitialization / deallocation prior to `gc_deinit` leaves the SSL object in an inconsistent state.

Rather than resolve the root cause, instead ensure that the closing of the user socket also closes the SSL socket.

Testing performed: on Feather ESP32S3 TFT run the reproducer script >10 times. Before, it crashed after 1 time. However, running 10 times helps show I'm not using up socket file descriptors accidentally.

This feels a bit voodoo so I'd love it if instead of incorporating this blindly, someone else went deeper and found the truth.

Closes: #6502

### CODE REVIEW

**Comment by dhalbert** [comment]
I am going to spend a little time seeing if I can find the root cause.

**Comment by dhalbert** [comment]
I do not understand the flow all that well, but this seems a bit fishy to me:
https://github.com/adafruit/circuitpython/blob/9e104c04aec3178b45ba70a302ad5800afad752f/ports/espressif/common-hal/socketpool/Socket.c#L609-L616
Under what circumstances is the passed in `socket` _not_ a `Socket` object. Could it be an `SSLSocket`? If so then fields in it are set willy-nilly. Also the SSLSocket is not shut down. And if it's a plain `Socket`, it's kind of being rendered unfunctional, but the underlying lwip socket isn't decommissioned in any way.

**Review by dhalbert — CHANGES_REQUESTED** [review]
One idea, and also a question in the main comments thread.

**Comment by jepler** [comment]
I had a theory that for an ssl socket `sss` created from a regular socket `s`, closing `s` first then `sss` could provoke the same problem without going through interpreter shutdown. However, I was not able to make this happen in practice.

**Comment by jepler** [comment]
The program I tried:
```
import wifi, socketpool, ssl, time, gc
wifi.radio.connect('<omitted>')
import socketpool
socket = socketpool.SocketPool(wifi.radio)
ctx = ssl.create_default_context()

b = bytearray(24)
for i in range(8):
    print(i)
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sss = ctx.wrap_socket(s)
    sss.connect(('finance.yahoo.com', 443))
    print(sss.send(b"GET /quote/%5EIXIC HTTP/1.0\r\nHost: finance.yahoo.com\r\n\r\n"))
    print(sss.recv_into(b))
    print(bytes(b))
    s.close()
    gc.collect()
    sss.close()
    gc.collect()
```

**Comment by jepler** [comment]
> Under what circumstances is the passed in `socket` _not_ a `Socket` object? Could it be an `SSLSocket`? If so then fields in it are set willy-nilly. Also the `SSLSocket` would not be shut down. And if it's a plain `Socket`, it's kind of being rendered unfunctional, but the underlying lwip socket isn't decommissioned in any way.

Scott answered this in a voice chat. The routine is actually _initializing_ a socket object for use by the web workflow. The name is not very self-documenting, so changing it and adding appropriate comments would be welcome as a separate PR.

File: ports/espressif/common-hal/socketpool/Socket.h [hunk 1]
```diff
@@ -42,6 +42,7 @@ typedef struct {
     int ipproto;
     bool connected;
     socketpool_socketpool_obj_t *pool;
+    struct ssl_sslsocket_obj *ssl_socket;
```
[the remark this answers is not in the document]
Comment (jepler) [reply ports/espressif/common-hal/socketpool/Socket.h]: Done

**Review by dhalbert — APPROVED** [review]
:+1:

**Pull request merged**